be/src/storage/rowset/json_column_iterator.cpp

```diff
@@ -86,14 +108,63 @@ Status JsonFlatColumnIterator::init(const ColumnIteratorOptions& opts) {
         RETURN_IF_ERROR(iter->init(opts));
     }
 
-    //  update stats
+    // update stats
     DCHECK(_path != nullptr);
     auto abs_path = _path->absolute_path();
     if (opts.stats->flat_json_hits.count(abs_path) == 0) {
         opts.stats->flat_json_hits[abs_path] = 1;
     } else {
         opts.stats->flat_json_hits[abs_path] = opts.stats->flat_json_hits[abs_path] + 1;
     }
+
+    DCHECK(_target_types.size() == _source_types.size());
+
+    for (int i = 0; i < _target_types.size(); i++) {
+        if (_target_types[i] == _source_types[i]) {
+            _cast_exprs.push_back(nullptr);
+            continue;
+        }
+
+        TypeDescriptor source_type(_source_types[i]);
+        TypeDescriptor target_type(_target_types[i]);
+
+        SlotDescriptor source_slot(i, "mock_solt", source_type);
+        ColumnRef* col_ref = _pool.add(new ColumnRef(&source_slot));
+
+        auto cast_expr = VectorizedCastExprFactory::from_type(source_type, target_type, col_ref, &_pool);
+        _cast_exprs.push_back(cast_expr);
+    }
+
+    return Status::OK();
+}
+
+template <typename FUNC>
+Status JsonFlatColumnIterator::_read_and_cast(JsonColumn* json_column, FUNC read_fn) {
+    json_column->init_flat_columns(_flat_paths, _target_types);
+    Chunk chunk;
+
+    for (int i = 0; i < _flat_iters.size(); i++) {
+        if (_cast_exprs[i] != nullptr) {
+            auto source = ColumnHelper::create_column(TypeDescriptor(_source_types[i]), true);
+            RETURN_IF_ERROR(read_fn(i, source.get()));
+
+            chunk.append_column(source, i);
+            ASSIGN_OR_RETURN(auto res, _cast_exprs[i]->evaluate_checked(nullptr, &chunk));
+            auto target = json_column->get_flat_field(i);
+            target->set_delete_state(source->delete_state());
+            if (res->only_null()) {
+                target->append_nulls(source->size());
+            } else if (res->is_constant()) {
+                auto data = down_cast<ConstColumn*>(res.get())->data_column();
+                target->append_value_multiple_times(*data, 0, source->size());
+            } else {
+                target->append(*res, 0, source->size());
```

Comment:
Is it possible that the target and res nullables don't match?

Reply:
I think never, source is a empty column, we can add a dcheck to do that